message_report: Improve message report template

PieterCK opened the pull request:
This updates the template structure of message report. 
- Simplified the template as per #36492.
- Excluded the message senders name on the list of recipients in GDM report.
- Added the message recipient's name in 1-on-1 DM report.
- Switched to displaying `reason_type`'s user-facing string.

**Screenshots**
(Updated Nov 10)

- Channel message reports:

  | Before | After |
  |--------|--------|
  | <img width="1075" height="407" alt="image" src="https://github.com/user-attachments/assets/7adcab77-2b4a-4635-855e-5a39f71efac0" /> | other's message <img width="1075" height="343" alt="image" src="https://github.com/user-attachments/assets/f4b7b512-cb98-4199-8ffc-f07b02450f51" /> |
  | <img width="1075" height="317" alt="image" src="https://github.com/user-attachments/assets/4872ed25-3aba-4896-9bb2-c88619696ed2" /> | one's own message <img width="1075" height="252" alt="image" src="https://github.com/user-attachments/assets/eaa7257e-049b-4d4d-8045-72601b1c5d2d" /> |

- Group direct message reports:

  | Before | After |
  |--------|--------|
  | <img width="1075" height="379" alt="image" src="https://github.com/user-attachments/assets/ffd9c332-c92f-4766-9b6d-0941fac29b44" /> | <img width="1075" height="312" alt="image" src="https://github.com/user-attachments/assets/c47beab4-2dfc-4fbb-a7c1-9f9eab0cb693" /> |
  | other's message <img width="1075" height="317" alt="image" src="https://github.com/user-attachments/assets/cbe2bab5-c1af-4a74-9386-7b34d708bbb3" /> | one's own message <img width="1075" height="252" alt="image" src="https://github.com/user-attachments/assets/ef4ea2dd-30cd-455c-b07d-38a791f56542" /> |
  | <img width="1075" height="263" alt="image" src="https://github.com/user-attachments/assets/d9941d86-b3e1-4d0e-bc9f-b48076f61501" /> | only two other recipients <img width="1075" height="250" alt="image" src="https://github.com/user-attachments/assets/749ae2a6-663c-49c4-9496-d0d881fa607f" /> |

- 1-on-1 direct message reports:

  | Before | After |
  |--------|--------|
  | <img width="1075" height="287" alt="image" src="https://github.com/user-attachments/assets/7849089b-70da-45f3-b9cc-63811841ffd9" /> |other's dm <img width="1075" height="220" alt="image" src="https://github.com/user-attachments/assets/d5af5294-2a8f-448f-a019-b792600f82ba" /> |
  | <img width="1075" height="379" alt="image" src="https://github.com/user-attachments/assets/11d71ee5-d129-482c-9d27-1e4e615100a4" /> |one's own dm to other <img width="1075" height="312" alt="image" src="https://github.com/user-attachments/assets/d6c0249d-069d-4504-b9f0-a06203e9d560" /> | 
  | <img width="1075" height="285" alt="image" src="https://github.com/user-attachments/assets/2e3a78d2-8fff-4bec-909e-dbe92eb77f58" /> |one's own dm to self <img width="1075" height="220" alt="image" src="https://github.com/user-attachments/assets/447ba31d-dab6-431b-a16e-2f6d2d5d2c24" /> | 

- With description:

  | Before | After |
  |--------|--------|
  | <img width="1075" height="409" alt="image" src="https://github.com/user-attachments/assets/25b9ba7c-e6bf-44cf-a8b2-6adec16ba38a" /> | <img width="1075" height="342" alt="image" src="https://github.com/user-attachments/assets/10a6b500-54bc-406d-bfbe-829363813881" /> |

Fixes: #36492


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented on zerver/lib/message_report.py:
```diff
@@ -38,11 +38,14 @@ def send_message_report(
             reporting_user_mention=reporting_user_mention,
             reported_user_mention=reported_user_mention,
         )
+        original_message_string = _("**Original message**").format(
```
Is just "Original message" OK for dm and gdm?

PieterCK replied:
Ok! Added message links for DMs and GDMs "Original message" string.

Diff https://github.com/zulip/zulip/pull/36498#issuecomment-3501516190

PieterCK commented on zerver/lib/message_report.py:
````diff
@@ -70,36 +76,33 @@ def send_message_report(
+**{report_type}**. {description}
+```
+
+{fence} spoiler {original_message_string}
````
Should be OK as one block, right? Since this doesn't have any translatable strings anymore.

PieterCK commented:
Empty quote block showing up as just ">" looks new to me, but most likely it just slipped my notice, since old PR for the message report API doesn't have an example where "notes" is empty. It is [reproducible](https://chat.zulip.org/#narrow/channel/7-test-here/topic/pieter/near/2293564)[#test here > pieter @ 💬](https://chat.zulip.org/#narrow/channel/7-test-here/topic/pieter/near/2293564) in CZO too.


````
```quote

```
````

PieterCK commented:
This is ready for review

timabbott commented on zerver/lib/message_report.py:
```diff
@@ -48,11 +48,13 @@ def send_message_report(
+        recipient_mentions: str = ", ".join(user_mentions)
         report_header = _(
             "{reporting_user_mention} reported a DM sent by {reported_user_mention} to {recipient_mentions}, and {last_user_mention}. Details:"
         ).format(
```
`recipient_users` would be a better variable name than "mentions" here for communicating with translators. Also, "direct message" rather than "DM" is what we use in user-facing strings.

PieterCK replied:
Changed "DM" → "direct message", updated the variable names, and logic for recipient user mentions is changed because https://github.com/zulip/zulip/pull/36498/commits/f28a869e640eed8cfc8d176cd1508e889c0509d9#r2502212869.

```diff
@@ -39,33 +40,26 @@ def send_message_report(
     # Build reported message header
     if is_1_to_1_message(reported_message):
         report_header = _(
-            "{reporting_user_mention} reported a DM sent by {reported_user_mention}. Details:"
+            "{reporting_user_mention} reported a direct message sent by {reported_user_mention}. Details:"
         ).format(
             reporting_user_mention=reporting_user_mention,
             reported_user_mention=reported_user_mention,
         )
-        original_message_string = _("**Original message**").format(
-            reported_user_mention=reported_user_mention
-        )
     elif reported_message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
-        recipient_list = get_display_recipient(reported_message.recipient)
-        user_mentions: list[str] = [
+        recipient_list = [
             silent_mention_syntax_for_user(user)
-            for user in recipient_list
+            for user in get_display_recipient(reported_message.recipient)
             if user["id"] is not reported_user.id
         ]
-        last_user_mention = user_mentions.pop()
-        recipient_mentions: str = ", ".join(user_mentions)
+        last_recipient_user = recipient_list.pop()
+        recipient_users: str = ", ".join(recipient_list)
         report_header = _(
-            "{reporting_user_mention} reported a DM sent by {reported_user_mention} to {recipient_mentions}, and {last_user_mention}. Details:"
+            "{reporting_user_mention} reported a direct message sent by {reported_user_mention} to {recipient_users}, and {last_recipient_user}. De
tails:"
         ).format(
             reporting_user_mention=reporting_user_mention,
             reported_user_mention=reported_user_mention,
-            recipient_mentions=recipient_mentions,
-            last_user_mention=last_user_mention,
-        )

```

PieterCK replied:
Should the  `reported/reporting_user_mention` variables also be renamed to just `reported/reporting_users`?

PieterCK commented on zerver/lib/message_report.py:
```diff
@@ -38,25 +39,25 @@ def send_message_report(
         )
     elif reported_message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
-        recipient_list = get_display_recipient(reported_message.recipient)
-        last_user = recipient_list.pop()
```
This is changed because `pop()`-ing the result of `get_display_recipient` here seems to mutates its return value for the same message if the function is called again.

alya commented:
We should explain where the message was sent in the odd "1 on 1 / to self" case. To avoid extra translations for this edge case, maybe

> .... sent by Desdemona to Desdemona.

alya commented:
Something is going wrong with how this list is being generated -- there's an extraneous comma.
<img width="300" height="39" alt="Screenshot 2025-11-07 at 09 26 04" src="https://github.com/user-attachments/assets/6ff86408-2ce7-4bb4-b1ef-2690ca906e00" />

PieterCK commented on zerver/lib/message_report.py:
```diff
@@ -70,36 +96,46 @@ def send_message_report(
+            realm,
+            dict(
+                id=reported_message.id,
+                display_recipient=get_display_recipient(reported_message.recipient),
```
`get_display_recipient` doesn't include the message sender's user data if it's a 1-1 message, so narrow URL for 1-1 DM doesn't work even if one of the recipient is the one who's trying to access it. Not sure if it's a bug or not.

timabbott replied:
That's expected, and hard to change, since `display_recipient` is in the API. We will need to separately work on how these URLs are handled.

PieterCK commented:
Marking this as ready for review! Updates since November 6:
- Switched "reported a DM sent" strings → "reported a direct message sent" 
- Moved the `original_message_string` logics to a separate block and added message link to "Original message" for private message reports.
- Tweaked the logic for recipient user mentions because of https://github.com/zulip/zulip/pull/36498#discussion_r2502212869.
- Renamed `recipient_mentions` and `last_user_mention` → `recipient_users` and `last_recipient_user`.
- Addressed Alya's [comments](https://github.com/zulip/zulip/pull/36498#issuecomment-3510330743) around some edge cases.
<details>

<summary>
git diff from November 6 (The version Tim did an initial review on)
</summary>

`````diff
diff --git a/zerver/lib/message_report.py b/zerver/lib/message_report.py
index a319e3fb89..8ccff1cbcf 100644
--- a/zerver/lib/message_report.py
+++ b/zerver/lib/message_report.py
@@ -7,6 +7,7 @@ from zerver.lib.markdown.fenced_code import get_unused_fence
 from zerver.lib.mention import silent_mention_syntax_for_user
 from zerver.lib.message import is_1_to_1_message, truncate_content
 from zerver.lib.topic_link_util import get_message_link_syntax
+from zerver.lib.url_encoding import pm_message_url
 from zerver.models import Message, Realm, UserProfile
 from zerver.models.recipients import Recipient
 from zerver.models.streams import StreamTopicsPolicyEnum
@@ -38,34 +39,50 @@ def send_message_report(
 
     # Build reported message header
     if is_1_to_1_message(reported_message):
-        report_header = _(
-            "{reporting_user_mention} reported a DM sent by {reported_user_mention}. Details:"
-        ).format(
-            reporting_user_mention=reporting_user_mention,
-            reported_user_mention=reported_user_mention,
-        )
-        original_message_string = _("**Original message**").format(
-            reported_user_mention=reported_user_mention
-        )
+        if reported_user != reporting_user:
+            report_header = _(
+                "{reporting_user_mention} reported a direct message sent by {reported_user_mention}."
+            ).format(
+                reporting_user_mention=reporting_user_mention,
+                reported_user_mention=reported_user_mention,
+            )
+        else:
+            # Clearly mention the direct message recipient if someone is
+            # reporting their own message in a 1-on-1 direct message.
+            if reported_message.recipient.type_id != reporting_user.id:
+                recipient_user = next(
+                    silent_mention_syntax_for_user(user)
+                    for user in get_display_recipient(reported_message.recipient)
+                    if user["id"] != reporting_user.id
+                )
+            else:
+                recipient_user = reporting_user_mention
+            report_header = _(
+                "{reporting_user_mention} reported a direct message sent by {reported_user_mention} to {recipient_user}."
+            ).format(
+                reporting_user_mention=reporting_user_mention,
+                reported_user_mention=reported_user_mention,
+                recipient_user=recipient_user,
+            )
     elif reported_message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP:
-        recipient_list = get_display_recipient(reported_message.recipient)
-        user_mentions: list[str] = [
-            silent_mention_syntax_for_user(user)
-            for user in recipient_list
-            if user["id"] is not reported_user.id
-        ]
-        last_user_mention = user_mentions.pop()
-        recipient_mentions: str = ", ".join(user_mentions)
+        recipient_list = sorted(
+            [
+                silent_mention_syntax_for_user(user)
+                for user in get_display_recipient(reported_message.recipient)
+                if user["id"] is not reported_user.id
+            ]
+        )
+        last_recipient_user = recipient_list.pop()
+        recipient_users: str = ", ".join(recipient_list)
+        if len(recipient_list) > 1:
+            recipient_users += ","
         report_header = _(
-            "{reporting_user_mention} reported a DM sent by {reported_user_mention} to {recipient_mentions}, and {last_user_mention}. Details:"
+            "{reporting_user_mention} reported a direct message sent by {reported_user_mention} to {recipient_users} and {last_recipient_user}."
         ).format(
             reporting_user_mention=reporting_user_mention,
             reported_user_mention=reported_user_mention,
-            recipient_mentions=recipient_mentions,
-            last_user_mention=last_user_mention,
-        )
-        original_message_string = _("**Original message**").format(
-            reported_user_mention=reported_user_mention
+            recipient_users=recipient_users,
+            last_recipient_user=last_recipient_user,
         )
     else:
         assert reported_message.is_channel_message is True
@@ -79,18 +96,31 @@ def send_message_report(
             reported_message.id,
         )
         report_header = _(
-            "{reporting_user_mention} reported a message sent by {reported_user_mention}. Details:"
+            "{reporting_user_mention} reported a message sent by {reported_user_mention}."
         ).format(
             reporting_user_mention=reporting_user_mention,
             reported_user_mention=reported_user_mention,
         )
-        original_message_string = _("**Original message at {channel_message_link}**").format(
-            channel_message_link=channel_message_link
-        )
 
     content = report_header
 
     # Build report context and message preview block
+    if reported_message.is_channel_message:
+        original_message_string = _("**Original message at {channel_message_link}**").format(
+            channel_message_link=channel_message_link
+        )
+    else:
+        direct_message_link = pm_message_url(
+            realm,
+            dict(
+                id=reported_message.id,
+                display_recipient=get_display_recipient(reported_message.recipient),
+            ),
+        )
+        original_message_string = _("**[Original message]({direct_message_link})**").format(
+            direct_message_link=direct_message_link
+        )
+
     reported_message_content = truncate_content(
         reported_message.content, MAX_REPORT_MESSAGE_SNIPPET_LENGTH, "\n[message truncated]"
     )

`````
</details>

alya commented:
Updated screenshots look good to me!

timabbott commented on zerver/lib/message_report.py:
```diff
@@ -34,29 +35,54 @@ def send_message_report(
+            else:
+                recipient_user = reporting_user_mention
+            report_header = _(
+                "{reporting_user_mention} reported a direct message sent by {reported_user_mention} to {recipient_user}."
```
Seems OK.

timabbott commented:
Merged, thanks @PieterCK!

We should make sure we have good documentation for the new feature.
